lib/validation/schema.ex

```diff
@@ -0,0 +1,25 @@
+defmodule Validation.Schema do
+  @moduledoc """
+  A Schema is a collection of rules that each are applied to the given params.
+  The result of evaluating a params map against a schema is a %Result{} struct.
+  """
+
+  alias Validation.Result
+
+  defstruct [
+    val: nil,
+    meta: []
+  ]
+
+  def build(rules) do
+    val = fn params ->
+      result = %Result{data: params}
+      rules
+      |> Enum.reduce(result, fn rule, result ->
```

Comment:
I think this could be simplified a bit. I would suggest to alter the Interface of a `Validation.Rule` a tiny bit.
I propose to add the following: 

```elixir
 def apply(%__MODULE__{ val: val },  result), do: val.(result)
```

This had the benefit of hiding the detail, that rules are implemented as compiled functions and provides a more generic interface that just depends on the module's struct.

With this function the code above can become:

```elixir
alias Validation.Rule
Enum.reduce(rules, result, &Rule.apply/2) 
```

This suggestion has the downside of having roughly one additional reduction due to the indirection of `apply/2`.
I think though that the advantage could justify it. (As an aside I think the apply interface could also be added to `Validation.Schema`)

What do you think?